File: src-tauri/src/game/dao_origin.rs
//! 天道本源：击键、读条、完美与出剑判定的唯一解释权。
//! 不负责 UI、货币兑换(iIoO)、拍场键位。

use std::time::Instant;

/// 关键帧窗口（读条进度比例）
pub const CRIT_LO: f64 = 0.76;
pub const CRIT_HI: f64 = 0.88;

/// 一次击锤的天道宣判（给 strike 用）
#[derive(Debug, Clone, Copy)]
pub struct StrikePulse {
    pub is_perfect: bool,
    /// 写入进度的倍率（完美 2.0，普通 1.0）
    pub progress_mul: f64,
}

/// 出剑前的天道宣判（给 SwordGenerator 用）
#[derive(Debug, Clone, Copy)]
pub struct ForgeVerdict {
    pub fail_rate: f64,
    pub drop_bonus: f64,
    pub rank_boost: u32,
    pub all_perfect: bool,
    pub perfect_hits: f64,
}

/// 天道本源：本轮读条 + 本剑熵
#[derive(Debug, Clone)]
pub struct DaoOrigin {
    pub cycle_start: Instant,
    /// 本剑完美累计（可与 GameState.forge_qte_hits 同步）
    pub perfect_hits: f64,
    pub strike_count: u32,
}

impl Default for DaoOrigin {
    fn default() -> Self {
        Self::new()
    }
}

impl DaoOrigin {
    pub fn new() -> Self {
        Self {
            cycle_start: Instant::now(),
            perfect_hits: 0.0,
            strike_count: 0,
        }
    }

    pub fn reset_cycle(&mut self) {
        self.cycle_start = Instant::now();
    }

    pub fn reset_sword(&mut self) {
        self.perfect_hits = 0.0;
        self.strike_count = 0;
        self.reset_cycle();
    }

    /// 当前读条进度 0~1
    pub fn progress(&self, interval_secs: f64) -> f64 {
        let secs = interval_secs.max(0.001);
        (self.cycle_start.elapsed().as_secs_f64() / secs).min(1.0)
    }

    pub fn in_crit_window(&self, interval_secs: f64) -> bool {
        let p = self.progress(interval_secs);
        p >= CRIT_LO && p < CRIT_HI
    }

    /// 手动空格或挂机满条
    pub fn on_strike(&mut self, manual: bool, power: f64) -> StrikePulse {
        let interval_guess = 1.0; // 仅用于帧内判定时由外部传入更准；保留 API
        let _ = interval_guess;
        self.strike_count = self.strike_count.saturating_add(1);

        if manual {
            // 帧内判定由调用方传入 is_perfect 更清晰——见 on_strike_crit
        }
        let _ = power;
        StrikePulse {
            is_perfect: false,
            progress_mul: 1.0,
        }
    }

    /// 推荐：调用方已算好是否帧内
    pub fn on_strike_with_crit(&mut self, in_crit: bool, power: f64) -> StrikePulse {
        self.strike_count = self.strike_count.saturating_add(1);
        if in_crit {
            self.perfect_hits += power;
            StrikePulse {
                is_perfect: true,
                progress_mul: 2.0,
            }
        } else {
            StrikePulse {
                is_perfect: false,
                progress_mul: 1.0,
            }
        }
    }

    /// 出剑结算：失败率 / 掉宝 / 品阶
    pub fn verdict_for_forge(&self, max_strikes: u32, base_god_rate: f64) -> ForgeVerdict {
        let hits = self.perfect_hits;
        let hits_u = hits.floor() as u32;
        let all_perfect = max_strikes > 0 && hits >= max_strikes as f64;

        let fail_rate = if all_perfect {
            0.0
        } else if hits_u >= 60 {
            0.01
        } else if hits_u >= 50 {
            0.02
        } else if hits_u >= 40 {
            0.05
        } else if hits_u > 30 {
            0.10
        } else {
            0.28
        };

        // 每次完美按威力折合：约 +2% 掉宝（与基线一致）
        let drop_bonus = (hits * 0.02).min(0.60);
        let rank_boost = hits_u.min(30);

        let _ = base_god_rate; // 合成在 state.effective_god_rate

        ForgeVerdict {
            fail_rate,
            drop_bonus,
            rank_boost,
            all_perfect,
            perfect_hits: hits,
        }
    }

    /// 与 state.forge_qte_hits 对齐时用
    pub fn sync_perfect_from_state(&mut self, hits: f64) {
        self.perfect_hits = hits;
    }
}

File: server/dao_origin.ts
// server/dao_origin.ts

export const CRIT_LO = 0.76;
export const CRIT_HI = 0.88;

export interface StrikePulse {
  is_perfect: boolean;
  progress_mul: number;
}

export interface ForgeVerdict {
  fail_rate: number;
  drop_bonus: number;
  rank_boost: number;
  all_perfect: boolean;
  perfect_hits: number;
}

export class DaoOrigin {
  public cycle_start: number;
  public perfect_hits: number;
  public strike_count: number;

  constructor() {
    this.cycle_start = Date.now();
    this.perfect_hits = 0;
    this.strike_count = 0;
  }

  public reset_cycle(): void {
    this.cycle_start = Date.now();
  }

  public reset_sword(): void {
    this.perfect_hits = 0;
    this.strike_count = 0;
    this.reset_cycle();
  }

  public progress(intervalSecs: number): number {
    const secs = Math.max(0.001, intervalSecs);
    const elapsedSecs = (Date.now() - this.cycle_start) / 1000.0;
    return Math.min(1.0, elapsedSecs / secs);
  }

  public in_crit_window(intervalSecs: number): boolean {
    const p = this.progress(intervalSecs);
    return p >= CRIT_LO && p < CRIT_HI;
  }

  public on_strike_with_crit(in_crit: boolean, power: number): StrikePulse {
    this.strike_count += 1;
    if (in_crit) {
      this.perfect_hits += power;
      return {
        is_perfect: true,
        progress_mul: 2.0,
      };
    } else {
      return {
        is_perfect: false,
        progress_mul: 1.0,
      };
    }
  }

  public verdict_for_forge(max_strikes: number, _base_god_rate: number): ForgeVerdict {
    const hits = this.perfect_hits;
    const hits_u = Math.floor(hits);
    const all_perfect = max_strikes > 0 && hits >= max_strikes;

    let fail_rate: number;
    if (all_perfect) {
      fail_rate = 0.0;
    } else if (hits_u >= 60) {
      fail_rate = 0.01;
    } else if (hits_u >= 50) {
      fail_rate = 0.02;
    } else if (hits_u >= 40) {
      fail_rate = 0.05;
    } else if (hits_u > 30) {
      fail_rate = 0.1;
    } else {
      fail_rate = 0.28;
    }

    const drop_bonus = Math.min(0.6, hits * 0.02);
    const rank_boost = Math.min(30, hits_u);

    return {
      fail_rate,
      drop_bonus,
      rank_boost,
      all_perfect,
      perfect_hits: hits,
    };
  }

  public sync_perfect_from_state(hits: number): void {
    this.perfect_hits = hits;
  }
}
